File: Managed/NextTurn.UE.Runtime/CoreUObject/BooleanProperty.cs
// Copyright (c) NextTurn. All rights reserved.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.txt in the project root for more information.

using System;
using NextTurn.UE;
using NextTurn.UE.Annotations;

namespace Unreal
{
    public sealed class BooleanProperty : Property, IProperty<bool>
    {
        internal BooleanProperty(IntPtr pointer) : base(pointer) { }

        public bool GetValue(Object @object, int index = 0)
        {
            if ((uint)index >= (uint)this.ArrayLength)
            {
                Throw.IndexArgumentOutOfRangeException();
            }

            return NativeMethods.GetValue(this.pointer, Object.GetPointerOrThrow(@object), index);
        }

        public void SetValue(Object @object, bool value, int index = 0)
        {
            if ((uint)index >= (uint)this.ArrayLength)
            {
                Throw.IndexArgumentOutOfRangeException();
            }

            NativeMethods.SetValue(this.pointer, Object.GetPointerOrThrow(@object), value, index);
        }

        private static new class NativeMethods
        {
            [Calli]
            public static extern bool GetValue(IntPtr property, IntPtr @object, int index);

            [Calli]
            public static extern void SetValue(IntPtr property, IntPtr @object, bool value, int index);
        }
    }
}

File: Source/NextTurnRuntime/Private/CoreUObject/BooleanProperty.cpp
// Copyright (c) NextTurn. All rights reserved.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.txt in the project root for more information.

#include "NextTurnRuntime/Private/Exports.hpp"

#include "NextTurnRuntime/Public/Compiler.hpp"

#include "Runtime/Core/Public/CoreMinimal.h"

#include "Runtime/CoreUObject/Public/UObject/UnrealType.h"

ENABLE_WARNINGS;

#define TYPE_EXPORT_MANAGED_TYPE_NAME BooleanProperty

TYPE_EXPORT_START

bool EXPORT_CALL_CONV GetValue(const FBoolProperty& Property, const UObject* Object, int32_t Index)
{
	return Property.GetPropertyValue_InContainer(Object, Index);
}

void EXPORT_CALL_CONV SetValue(const FBoolProperty& Property, UObject* Object, bool Value, int32_t Index)
{
	Property.SetPropertyValue_InContainer(Object, Value, Index);
}

const FMemberSymbol Members[] =
{
	EXPORT_METHOD(GetValue),
	EXPORT_METHOD(SetValue),
};

TYPE_EXPORT_END

#undef TYPE_EXPORT_MANAGED_TYPE_NAME

RESTORE_WARNINGS;
